common/reactive/src/main/java/io/helidon/common/reactive/MultiFromInputStream.java

```diff
@@ -0,0 +1,145 @@
+
+    private final InputStream inputStream;
+    private final int bufferSize;
+
+    MultiFromInputStream(InputStream inputStream, int bufferSize) {
+        this.inputStream = inputStream;
+        this.bufferSize = bufferSize;
+    }
+
+    @Override
+    public void subscribe(final Flow.Subscriber<? super ByteBuffer> subscriber) {
+        Objects.requireNonNull(subscriber, "subscriber is null");
+        try {
+            inputStream.available();
+        } catch (IOException e) {
+            subscriber.onSubscribe(EmptySubscription.INSTANCE);
+            subscriber.onError(e);
+            return;
+        }
+        InputStreamSubscription subscription = new InputStreamSubscription(subscriber, inputStream, bufferSize);
+        subscriber.onSubscribe(subscription);
+    }
+
+    static final class InputStreamSubscription extends AtomicLong implements Flow.Subscription {
+
+        private final Flow.Subscriber<? super ByteBuffer> downstream;
+
+        private InputStream inputStream;
+        private final int bufferSize;
+
+        private volatile int canceled;
+
+        static final int NORMAL_CANCEL = 1;
+        static final int BAD_REQUEST = 2;
+
+        InputStreamSubscription(Flow.Subscriber<? super ByteBuffer> downstream, InputStream inputStream, final int bufferSize) {
+            this.downstream = downstream;
+            this.inputStream = inputStream;
+            this.bufferSize = bufferSize;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0L) {
+                canceled = BAD_REQUEST;
+                n = 1; // for cleanup
+            }
+
+            if (SubscriptionHelper.addRequest(this, n) != 0L) {
+                return;
+            }
+
+            long emitted = 0L;
+            Flow.Subscriber<? super ByteBuffer> downstream = this.downstream;
+
+            for (;;) {
+                while (emitted != n) {
+                    int isCanceled = canceled;
+                    if (isCanceled != 0) {
+                        var toClose = inputStream;
+                        inputStream = null;
+                        if (isCanceled == BAD_REQUEST) {
+                            downstream.onError(new IllegalArgumentException(
+                                    "Rule §3.9 violated: non-positive request amount is forbidden"));
+                        } else if (isCanceled == NORMAL_CANCEL) {
+                            try {
+                                toClose.close();
+                            } catch (IOException e) {
+                                // Rule §1.8 says eventually
+                                downstream.onError(e);
+                            }
+                        }
+                        return;
+                    }
+
+                    ByteBuffer value;
+
+                    try {
+                        value = ByteBuffer.wrap(inputStream.readNBytes(bufferSize));
```

Comment:
AFAIK this method should not be blocking. If you block here, you block the subscriber trying to request data.
It feels the only way to implement this correctly is to use an executor service.
I think the existing implementations are not correct.

Reply:
Depends if you can trust InputStream to not block, then it would be excusable I guess, after all every method out of our control can block